### PR TITLE
feat: Webpack 5 Experimental Support

### PR DESCRIPTION
**Pull request by ScriptedAlchemy**
This is a WIP PR to add Webpack 5 support to next.js

There are some outstanding areas that will need work. But it's functional. 

- [x] `on-demand-entries-handler` needs to be rewritten and provide WP4 & WP5 compatibility
- [ ] `formatMessages` needs to be updated to parse output messages again, most of it is commented out right now
- [x] Solution to selectively install WP5 or WP4 along with upgrading dev-middleware
- [ ] `EsmPlugin` - Needs a dependency factory because MiltiEntryPlugin was removed 
- [ ] TypeScript definitions and types need some updates
- [x] require cache is messed up in dev mode, likely due to `on-demand-entries-handler` - a workaround I'm using is `uncachedRequire` 
- [x]  `Module not found: Error: Can't resolve 'pnpapi' in '/next.js/node_modules/enhanced-resolve/lib'"` When doing a seerverless build
- [x] cleanup `hot-reloader`
- [x] API routes are not found. 

This pull request is also needed.
https://github.com/zeit/next-plugins/pull/620

I am publishing all this code to npm under module-federation org https://github.com/module-federation

### CODE REVIEW

File: packages/next/build/index.ts [hunk 1]
```diff
@@ -369,9 +369,10 @@ export default async function build(dir: string, conf = null): Promise<void> {
   if (buildSpinner) {
     buildSpinner.stopAndPersist()
   }
-  console.log()
-
-  result = formatWebpackMessages(result)
+  console.log('There were errors!')
```
Comment (ScriptedAlchemy) [inline packages/next/build/index.ts]: needs to be fixed

File: packages/next/build/webpack-config.ts [hunk 2]
```diff
@@ -34,14 +34,13 @@ import ChunkNamesPlugin from './webpack/plugins/chunk-names-plugin'
 import { CssMinimizerPlugin } from './webpack/plugins/css-minimizer-plugin'
 import { importAutoDllPlugin } from './webpack/plugins/dll-import'
 import { DropClientPage } from './webpack/plugins/next-drop-client-page-plugin'
-import NextEsmPlugin from './webpack/plugins/next-esm-plugin'
```
Comment (ScriptedAlchemy) [inline packages/next/build/webpack-config.ts]: broken in WP5, will address later

File: packages/next/build/webpack-config.ts [hunk 3]
```diff
@@ -777,41 +783,56 @@ export default async function getBaseWebpackConfig(
...
-                  config: {
-                    devtool,
-                    mode: webpackMode,
-                    resolve: resolveConfig,
-                  },
-                })
-              )
+              if (!webpack5Experiential) {
+                // Not needed with webpack 5
+                const AutoDllPlugin = importAutoDllPlugin({ distDir })
```
Comment (ScriptedAlchemy) [inline packages/next/build/webpack-config.ts]: WP5 has its own cache, and its faast

File: packages/next/client/router.ts [hunk 4]
```diff
@@ -120,7 +120,13 @@ export default singletonRouter as SingletonRouter
 export { default as withRouter } from './with-router'
 
 export function useRouter() {
-  return React.useContext(RouterContext)
+  if (process.browser) {
```
Comment (ScriptedAlchemy) [inline packages/next/client/router.ts:123]: Need some help on this one, HMR or SSG seems to lose context

File: packages/next/next-server/server/render.tsx [hunk 5]
```diff
@@ -442,6 +442,8 @@ export async function renderToHTML(
   const router = new ServerRouter(pathname, query, asPath, {
     isFallback: isFallback,
   })
+  global.router = router
```
Comment (ScriptedAlchemy) [inline packages/next/next-server/server/render.tsx:445]: another hack, losing context during HMR or SSG

**Comment by ScriptedAlchemy** [comment]
Against master, 1700 tests pass. Most failures are around HMR or SSG - I resolved many of these with workarounds. The remaining tests are mostly related to format messages

@timneutkens Willing and ready to work with your team on this upgrade. From my perspective, its pretty far along.

Canary introduces some changes which I'm not sure how to address, like externals for webpack-dev-server which no longer exist in v4 rc

**Comment by timneutkens** [comment]
Going to close this as I've built out the webpack 5 support, only thing left is some fast refresh tests that aren't passing.

Related PRs:
#13759
#13870
#13937
#14130
#14176
#14268
https://github.com/vercel/next.js/pull/14279
https://github.com/vercel/next.js/pull/14413
https://github.com/vercel/next.js/pull/14498
https://github.com/vercel/next.js/pull/14527
https://github.com/vercel/next.js/pull/14626
https://github.com/vercel/next.js/pull/14722
https://github.com/vercel/next.js/pull/14793

**Pull request closed without merging**